### ISSUE TITLE
Light sensor Kobo Aura One (feature request) 

### ISSUE THREAD

**Issue by wvhulle** [koreader/koreader#2500]
Hi all, 

I know there is work going on making the red LEDs work but the intensity of the white and red LEDs is controlled by a light sensor in nickel. Without running nickel this brightness level is not updated automatically. How could I help to make this possible in Koreader?

Best regards

**Comment by houqp** [comment]
We need to find out how to read the sensor output from the kernel, which might require a little bit of reverse engineering. One possible place to look at is /dev/input.

**Comment by wvhulle** [comment]
Is there an easy way to read files in /dev/input while the devices is connect over usb?

**Comment by houqp** [comment]
Yes, you should be able to still read all the input from /dev/input while device is connected to usb. I recall there is a builtin command to dump events from /dev/input, but I can't remember the name. You can just write a simple problem in C or LuaJIT to read and print the events: http://stackoverflow.com/questions/15949163/read-from-dev-input

**Comment by pazos** [comment]
@houqp: evtest, but seems broken on my KA1 (te program starts and enumerate input types, but doesn't generate any events on touch), it worked ok on my Aura HD. Could you check this?

**Comment by dengste** [comment]
So just judging from the names this could be a Si1145/46/47 ambient light sensor:

https://www.allaboutcircuits.com/uploads/articles/si1145-46-47.pdf

I couldn't find  a teardown of the KA1 to affirm this, so if anyone has ever disassembled his KA1 and made pics of the board, this would be helpful.

**Comment by Frenzie** [comment]
@dengste There's this: https://youtu.be/gT3VuGT0YyY?t=2m42s Perhaps it shows something relevant by accident. :-P

**Comment by dengste** [comment]
That video only shows the back, but the sensor is on the front. However, sometimes things are just too damn simple:

    cat /sys/devices/virtual/input/input3/name
      ---> si114x_als_vis

Also, /proc/config.gz has CONFIG_INPUT_SI114X=y. This is an upstream driver, so we should have no problems dealing with this device at all.

**Comment by poire-z** [comment]
> but I'm not sure how to set the frontlight in the background. Should I use the backgroundrunner plugin for this?

You could use backgroundrunner, but having something executes itself every N seconds is as simple as:
```lua
self.update_action = function()
    print("bouh") -- do you stuff
    if not self.should_stop_updating then
        UIManager:scheduleIn(20, self.update_action) -- have me been called again in 20s
    end
end
UIManager:scheduleIn(5, self.update_action) -- start it in 5s
-- elsewhere, if you want to cancel it:
UIManager:unschedule(self.update_action)

-- if you use a local for it, you need to define it before referencing it into function:
-- local update_action = nil
-- update_action = function()
--        UIManager:scheduleIn(20, update_action)
-- end
```

(There is a [autofrontlight plugin ](https://github.com/koreader/koreader/tree/master/plugins/autofrontlight.koplugin) that uses background runner).

But if by "in the background", you mean that your reading of /sys/ stuff or your writting to it may block the user for a few seconds, this is another kind of problem.

**Comment by poire-z** [comment]
Not sure when such a thing should be started, From the device or from a plugin ?

Anyway, in frontend/device, someone has met your problem and used this:
https://github.com/koreader/koreader/blob/680b53e3470b95a50686987439456a7bf7b1672d/frontend/device/generic/powerd.lua#L158-L167

**Comment by Frenzie** [comment]
@dengste It sounds like you should copy or adjust https://github.com/koreader/koreader/blob/680b53e3470b95a50686987439456a7bf7b1672d/plugins/autofrontlight.koplugin/main.lua which does something very similar.

**Comment by dengste** [comment]
I'm not sure where this should go either. I think the configuration for this should be part of the normal frontlightwidget, since I think otherwise users will be confused when the plugin overrides the settings there. This is why I started adding this stuff to kobo/powerd, so that the frontlightwidget as direct access to it. But indeed, it seems this is probably better put in a separate plugin. In the frontlightwidget, I could detect if the plugin is active and then add the necessary widgets to configure it. Does that sound reasonable?

**Comment by Frenzie** [comment]
In that case I think it would conceptually make more sense to do it from the plugin itself? Very crudely speaking, something like:

```lua
local FrontLightWidgetThingy = require "frontlightwidget"

FrontLightWidgetThingy.special_new_plugin_area = necessary widget frame stuff
-- or some kind of function?
FrontLightWidgetThingy:addWidget()
```

Basically to make it accommodate the principle but without some weird core/plugin circular dependency.

Alternatively just stick the initialization in UIManager:init()? In the if:isKobo() section Device.powerd.startAutoWarmthSchedule (or whatever). Or perhaps better yet, something like Device:initSchedulers in generic device, which can then be called from UIManager. We do want to stick that in core one of these days… https://github.com/koreader/koreader/pull/3078#issuecomment-324106748

**Comment by dengste** [comment]
Thanks for the suggestions. I like the idea of introducing placeholders in the frontlightwidget for plugins, so I think I'll go with that.

**Comment by lyntur** [comment]
Sorry for bumping this up but is there any news on the auto frontlight working on Kobo Aura One?

**Comment by NiLuJe** [comment]
That would require someone with the device actually looking into it ;o).
